pkg/apis/resource/structured/namedresources/validation/validation.go

```diff
@@ -0,0 +1,184 @@
+)
+
+type Options struct {
+	// StoredExpressions must be true if and only if validating CEL
+	// expressions that were already stored persistently. This makes
+	// validation more permissive by enabling CEL definitions that are not
+	// valid yet for new expressions.
+	StoredExpressions bool
+}
+
+func ValidateResources(resources *resource.NamedResourcesResources, fldPath *field.Path) field.ErrorList {
+	allErrs := validateInstances(resources.Instances, fldPath.Child("instances"))
+	return allErrs
+}
+
+func validateInstances(instances []resource.NamedResourcesInstance, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	instanceNames := sets.New[string]()
+	for i, instance := range instances {
+		idxPath := fldPath.Index(i)
+		instanceName := instance.Name
+		allErrs = append(allErrs, validateInstanceName(instanceName, idxPath.Child("name"))...)
+		if instanceNames.Has(instanceName) {
+			allErrs = append(allErrs, field.Duplicate(idxPath.Child("name"), instanceName))
+		} else {
+			instanceNames.Insert(instanceName)
+		}
+		allErrs = append(allErrs, validateAttributes(instance.Attributes, idxPath.Child("attributes"))...)
+	}
+	return allErrs
+}
+
+var (
+	numericIdentifier = `(0|[1-9]\d*)`
+
+	preReleaseIdentifier = `(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)`
+
+	buildIdentifier = `[0-9a-zA-Z-]+`
+
+	semverRe = regexp.MustCompile(`^` +
+
+		// dot-separated version segments (e.g. 1.2.3)
+		numericIdentifier + `\.` + numericIdentifier + `\.` + numericIdentifier +
+
+		// optional dot-separated prerelease segments (e.g. -alpha.PRERELEASE.1)
+		`(-` + preReleaseIdentifier + `(\.` + preReleaseIdentifier + `)*)?` +
+
+		// optional dot-separated build identifier segments (e.g. +build.id.20240305)
+		`(\+` + buildIdentifier + `(\.` + buildIdentifier + `)*)?` +
+
+		`$`)
+)
+
+func validateAttributes(attributes []resource.NamedResourcesAttribute, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	attributeNames := sets.New[string]()
+	for i, attribute := range attributes {
+		idxPath := fldPath.Index(i)
+		attributeName := attribute.Name
+		allErrs = append(allErrs, validateAttributeName(attributeName, idxPath.Child("name"))...)
+		if attributeNames.Has(attributeName) {
+			allErrs = append(allErrs, field.Duplicate(idxPath.Child("name"), attributeName))
+		} else {
+			attributeNames.Insert(attributeName)
+		}
+
+		entries := sets.New[string]()
+		if attribute.QuantityValue != nil {
+			entries.Insert("quantity")
+		}
+		if attribute.BoolValue != nil {
+			entries.Insert("bool")
+		}
+		if attribute.IntValue != nil {
+			entries.Insert("int")
+		}
+		if attribute.IntSliceValue != nil {
+			entries.Insert("intSlice")
+		}
+		if attribute.StringValue != nil {
+			entries.Insert("string")
+		}
+		if attribute.StringSliceValue != nil {
+			entries.Insert("stringSlice")
+		}
+		if attribute.VersionValue != nil {
+			entries.Insert("version")
+			if !semverRe.MatchString(*attribute.VersionValue) {
```

Comment:
Any particular reason why not just use `semver.Parse` here, instead of maintaining regexp? Don't get me wrong I like regex (I know I'm a weirdo). Still the question stands.

Reply:
Jordan was concerned that depending on a third-party lib for API validation might cause problems if the implementation of that lib changes. The regexp is entirely under our control.

Reply:
https://github.com/kubernetes/kubernetes/pull/123516#discussion_r1509348353

Reply:
Fair